README.md


# Screenly OSE Monitoring
<p align="center">
<img width="800px" title="Manage Monitoring" alt="Manage Monitoring" src="https://raw.githubusercontent.com/didiatworkz/screenly-ose-monitor/v3.0/.github/IMAGES/manage5.png" />
</p>

![GitHub release (latest by date)](https://img.shields.io/github/v/release/didiatworkz/screenly-ose-monitor) ![GitHub Release Date](https://img.shields.io/github/release-date/didiatworkz/screenly-ose-monitor?color=green) ![GitHub last commit (branch)](https://img.shields.io/github/last-commit/didiatworkz/screenly-ose-monitor/v3.0) ![GitHub commit activity](https://img.shields.io/github/commit-activity/y/didiatworkz/screenly-ose-monitor)  ![GitHub issues](https://img.shields.io/github/issues-raw/didiatworkz/screenly-ose-monitor)  ![GitHub stars](https://img.shields.io/github/stars/didiatworkz/screenly-ose-monitor?style=social)

- [Intro](#what-is-this)
- [Features](#features)
- [Bugs](#Bugs)
- [Requirements](#requirements)
- [Installation](#installation)
- [Login](#login)
- [Update](#update)
- [Changelog](#changelog)
- [Bash Controll](#bash-controll)
- [More Questions?](#any-more-questions)


## What is this?
<img align="left" width="300px" title="Monitoring Overview" alt="Monitoring Overview" src="https://raw.githubusercontent.com/didiatworkz/screenly-ose-monitor/v3.0/.github/IMAGES/manage4.png" style="padding-right: 12px; padding-bottom:12px" />

Screenly OSE Monitoring is a web-based application that is simultaneously installed on a Screenly OSE Player or as standalone solution.

With this tool, you can manage multiple OSE players in one interface.

In addition to its current display status, you can also manage the assets of a player. It is possible to activate or deactivate assets, add new ones, edit or simply delete them all from the player.
<br />
<br />
By installing an add-on on each player, it is also possible to display a "live feed" of the player's HDMI output. This will then be displayed in the overview, as well.




## Features

<img width="300px" align="right" src="https://raw.githubusercontent.com/didiatworkz/screenly-ose-monitor/v3.0/.github/IMAGES/monitoring2.png">

+ Easy administration
+ User management
+ Simple overview of all players with status
+ Add / Edit / Remove / Order / Control Assets
+ Auto discovery of players
+ Responsive Design
+ Public access to integrate in Monitoring tools
+ Add-on for displaying the player’s output remotely

## Bugs
No errors were found in the tests. If you find a problem or bug, please report it:<br />
[Open Issue List](https://github.com/didiatworkz/screenly-ose-monitor/issues?q=is:issue%20label:bug)

---

## Requirements
+ RaspberryPi 3B+
+ Raspbian Lite
+ PHP 7.x (will be installed)
+ SQLite 3.x (will be installed)
+ Ansible (will be installed)

## Installation
__IMPORTANT__: The monitoring was designed to run on the local network with the Screenly OSE Player. Problems or restrictions may occur if the server is hosted externally or accessed from an external network!

Very simple installation:

1. Connect to the player via SSH or terminal ([CTRL]+[ALT]+[F1])
2. Copy this line and execute it
```sh
bash <(curl -sL https://git.io/fjg7h)
```
3. Answer the questions and installation will be start (This may take a while - Don't worry)
4. Open your Browser to the IP address of the Raspberry Pi like: http://[screenly-ip-address]:9000

## Login
After the installation is the default login:

http://[screenly-ip-address]:9000

Username: demo<br />
Password: demo

<p align="center">
<img width="500px" title="Manage Monitoring" alt="Manage Monitoring" src="https://raw.githubusercontent.com/didiatworkz/screenly-ose-monitor/v3.0/.github/IMAGES/livecase.png" />
</p>

---

## Update
For the update you only have to run the installer again.
The installer checks if there is an old installation and saves it if necessary.
From version 2.0 there are changes to the database structure which is prepared for future updates and no longer needs to be saved.

But please note that the Add-on may have to be reinstalled on the players. (Current for all older versions before version 2.0)

__IMPORTANT: If monitoring is installed on a Raspberry Pi that does not have a Screenly OSE Player (standalone installation) the port changes from 9000 to 80.
So the monitoring is accessible via the normal IP address!__

### Monitoring
```sh
bash <(curl -sL https://git.io/fjg7h)
```

### Extensions
Link over Webfrontend or this:
```sh
bash <(curl -sL https://git.io/fjg5e)
```

### Changelog
[Open Changelog](https://github.com/didiatworkz/screenly-ose-monitor/blob/master/CHANGELOG.md)


<p align="center">
<img title="Monitoring Overview" alt="Monitoring Overview" src="https://raw.githubusercontent.com/didiatworkz/screenly-ose-monitor/v3.0/.github/IMAGES/screens2.png" />
</p>

## Bash Controll
Since version 2.0 there is a small possibility to update or check ose-monitoring via bash.
For more info check:
```sh
ose-monitoring --help
```

## Any more questions?
There's something else that wasn't answered here?
Then just have a look at the [wiki-page]([https://github.com/didiatworkz/screenly-ose-monitor/wiki](https://github.com/didiatworkz/screenly-ose-monitor/wiki)). Maybe you will find an answer there.

Thanks for using this project! <br />
\- didiatworkz

<p align="center">
<img width="128px" title="atworkz logo" alt="atworkz logo" src="https://assets.atworkz.de/img/atworkz_logo_512.png" />
</p>


CHANGELOG.MD
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [3.0] - 2020-05-20

### Added
- Setup Wizard
- Image and Video upload
- Multiuploader for upload files to multiple player
- User management
- Auto discovery for add players
- Clean assets function

### Changed
- Sort assets
- Improved instalation speed
- Player rotation for public link
- php.ini parameter
- monitoring.conf template
- screenshot image refresh
- Fix name bug
- Group Assets
- Modal display


## [2.4] - 2019-10-31

### Changed
- Dynamic php-fpm package installation
- Error when no Player is set by using the token
- start/end date/time not saved when you edit a asset

## [2.2] - 2019-10-21
### Added
- Reboot function
- Standalone installation

### Changed
- Installer problems
- Linked php7.0 -> php (7.3)
- Package names
- Layout problems
- Black-Dashboard Design bugs
- Screenshot add-on


## [2.1] - 2019-05-17
### Added
- Managing Assets
- Add Assets
- Edit Assets
- New Design
- Monitoring Token
- Bash Tool
- Screenshots are stored in the RAM

### Changed
- Performance
- Control Assets
- Screenshot add-on
- Ansible installer

## [1.2] - 2019-03-13

### Changed
- Script installer
- Improve Extension


_config.php
<?php
  /*
	                            _
	   ____                    | |
	  / __ \__      _____  _ __| | __ ____
	 / / _` \ \ /\ / / _ \| '__| |/ /|_  /
	| | (_| |\ V  V / (_) | |  |   <  / /
	 \ \__,_| \_/\_/ \___/|_|  |_|\_\/___|
	  \____/

			http://www.atworkz.de
			   info@atworkz.de
	________________________________________
			  Screenly OSE Monitor
		   Version 3.0 - November 2019
	________________________________________
	*/
	ini_set('display_errors',1);
	error_reporting(E_ALL|E_STRICT);

	$dbase_key		= 'assets/tools/key.php';
	if(!@file_exists($dbase_key)) {
		$dbase_file = 'dbase.db';
	} else {
		include_once($dbase_key);
		$dbase_file = $db_cryproKey;
		if(@file_exists('dbase.db')) {
			unlink('dbase.db');
		}

	}

	$db 			= new SQLite3($dbase_file);
	$set 			= $db->query("SELECT * FROM settings WHERE userID = 1");
	$set 			= $set->fetchArray(SQLITE3_ASSOC);
	$loginUsername 	= $set['username'];
	$loginPassword 	= $set['password'];
	$loginUserID 	= $set['userID'];
	$securityToken	= $set['token'];
	$updatecheck	= $set['updatecheck'];
	$systemVersion  = file_get_contents('assets/tools/version.txt');
	$apiVersion		= 'v1.2';

	if(!@file_exists($dbase_key)){
		$token = md5($systemVersion.time().$loginPassword).'.db';
		$keyFile = '<?php
		$db_cryproKey = "'.$token.'";';
		$current = file_get_contents($dbase_key);
		file_put_contents($dbase_key, $keyFile);
		rename("dbase.db",$token);
		header("Refresh:0");
		die("Reload this page");
	}

	if(@file_exists('assets/tools/version_old.txt')){
		$oldVersion = file_get_contents('assets/tools/version_old.txt');
		if($oldVersion <= '2.0'){			// Update Database to Version 2.0
			$db->exec("ALTER TABLE `settings` ADD COLUMN `token` TEXT");
			$db->exec("ALTER TABLE `settings` ADD COLUMN `end_date` INTEGER");
			$db->exec("ALTER TABLE `settings` ADD COLUMN `duration` INTEGER");
			$db->exec("UPDATE `settings` SET token='d1bf93299de1b68e6d382c893bf1215f' WHERE userID=1");
			$db->exec("UPDATE `settings` SET end_date=1 WHERE userID=1");
			$db->exec("UPDATE `settings` SET duration=30 WHERE userID=1");
		}
		if($oldVersion <= '2.1'){			// Update Database to Version 2.1
			$db->exec("ALTER TABLE `settings` ADD COLUMN `updatecheck` INTEGER");
			$db->exec("ALTER TABLE `settings` ADD COLUMN `refreshscreen` INTEGER");
			$db->exec("UPDATE `settings` SET updatecheck=0 WHERE userID=1");
			$db->exec("UPDATE `settings` SET refreshscreen=5 WHERE userID=1");
		}
		if($oldVersion <= '3.0'){			// Update Database to Version 3.0
			//Nothing
		}
		unlink('assets/tools/version_old.txt');
		unlink('update.txt');
		header("Refresh:0");
		die("Reload this page");
	}

	if(isset($_GET['site'])){
		$site = $_GET['site'];
	} else $site = NULL;

	function redirect($url, $time = 1){
		echo '<meta http-equiv="refresh" content="'.$time.';URL='.$url.'">';
	}

	function sysinfo($status, $message, $refresh = false){
		echo '<script>$.notify({icon: "tim-icons icon-bell-55",message: "'.$message.'"},{type: "'.$status.'",timer: 1000,placement: {from: "top",align: "center"}});</script>';
		if($refresh) echo'<meta http-equiv="refresh" content="2;URL=index.php">';
	}

	function playerAuthentication($value = null){
		global $dbase_file;
		$db = new SQLite3($dbase_file);
		if(filter_var($value, FILTER_VALIDATE_IP)){
			$playerSQL 	= $db->query("SELECT * FROM player WHERE address='".$value."'");
		}
		else if(is_numeric($value)){
			$playerSQL 	= $db->query("SELECT * FROM player WHERE playerID='".$value."'");
		}
		else return FALSE;
		$player 		= $playerSQL->fetchArray(SQLITE3_ASSOC);
		$player['player_user'] != '' ? $user = $player['player_user'] : $user = false;
		$player['player_password'] != '' ? $pass = $player['player_password'] : $pass = false;
		return array('username' => $user, 'password' => $pass);
	}

	function callURL($method, $ip, $params = false, $playerID = null, $ssl = false){
		$headers = array(
			'Accept: application/json',
			'Content-Type: application/json',
		);
		$curl = curl_init();
		if($ssl) $prefix = 'https://';
		else $prefix = 'http://';

		$playerAuth = playerAuthentication($playerID);
		if($playerAuth['username'] != '' AND $playerAuth['password'] !=''){
			$user = $playerAuth['username'];
			$pass = $playerAuth['password'];
		}
		else {
			$user = false;
			$pass = false;
		}
		if($user AND $pass) $url = $prefix.$user.':'.$pass.'@'.$ip;
		else $url = $prefix.$ip;

		switch($method){
			case 'GET':
				//$url .= '?' . http_build_query($params);
				break;
			case 'POST':
				curl_setopt($curl, CURLOPT_POST, true);
				curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($params));
				break;
			case 'PUT':
				curl_setopt($curl, CURLOPT_CUSTOMREQUEST, 'PUT');
				curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($params));
				break;
			case 'DELETE':
				curl_setopt($curl, CURLOPT_CUSTOMREQUEST, 'DELETE');
				//$url .= '?' . http_build_query($params);
				break;
		}

		curl_setopt($curl, CURLOPT_CONNECTTIMEOUT_MS, 250);
		curl_setopt($curl, CURLOPT_URL, $url);
		curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
		curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
		curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

		$response = curl_exec($curl);
		$code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
		curl_close($curl);

		if ($code == 200) {
			return $response = json_decode($response, true);
		}
		elseif ($code == 301) {
		   return callURL($method, $ip, $params, $playerID, true);
		}
		elseif ($code == 401) {
			sysinfo('warning', 'Can not logged in to the player! - Wrong Username or Password!');
			return 'authentication error '.$code;
		}
		else return 'error '.$code;
	}

	function checkAddress($ip){
		$ch = curl_init($ip);
		curl_setopt($ch, CURLOPT_TIMEOUT, 1);
		curl_setopt($ch, CURLOPT_CONNECTTIMEOUT_MS, 200);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
		$data = curl_exec($ch);
		$httpcode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
		curl_close($ch);
		if(($httpcode>=200 && $httpcode<300) || $httpcode==401) return true;
		else return false;
	}

	function monitorScript($url){
		if(checkAddress($url)) {
			$monitor = callURL('GET', $url.':9020/monitor.txt');
			if($monitor == 1) return 'http://'.$url.':9020/screen/screenshot.png';
			else return 'assets/img/online.png';
		}
		else return 'assets/img/offline.png';
	}


	if($updatecheck < time() && (date("d", $updatecheck) != date("d"))){
		shell_exec('ose-monitoring --scriptupdate');
		$db->exec("UPDATE `settings` SET updatecheck='".time()."' WHERE userID=1");
	}

	if(@file_exists('update.txt')) {
		$update = '
					<li class="nav-item">
						<a href="https://github.com/didiatworkz/screenly-ose-monitor" target="_blank" class="nav-link" data-tooltip="tooltip" data-placement="bottom" title="Update available!">
							<i class="tim-icons icon-cloud-download-93 blink"></i>
							<p class="d-lg-none">
								Update available!
							</p>
						</a>
					</li>
		';

	}
	else $update = '';

	if(isset($_POST['newAsset'])){
		$now				= strtotime("-10 minutes");
		$id 				= $_POST['id'];
		$url 				= $_POST['url'];
		$name 			= (isset($_POST['name']) ? $_POST['name'] : $_POST['url']);
		$mimetype		= $_POST['mimetype'];
		$start 			= date("Y-m-d", $now);
		$start_time	= "00:00";
		$end 				= date("Y-m-d", strtotime("+".$set['end_date']." week"));
		$end_time		= $start_time;
		$duration 	= $set['duration'];

		if($name == '') $name = $url;

		$playerSQL 	= $db->query("SELECT * FROM player WHERE playerID='".$id."'");
		$player 		= $playerSQL->fetchArray(SQLITE3_ASSOC);

		$data 										= array();
		$data['mimetype'] 				= $mimetype;
		$data['is_enabled'] 			= 1;
		$data['name'] 						= $name;
		$data['start_date'] 			= $start.'T'.$start_time.':00.000Z';
		$data['end_date'] 				= $end.'T'.$end_time.':00.000Z';
		$data['duration'] 				= $duration;
		$data['play_order']				= 0;
		$data['nocache'] 					= 0;
		$data['uri'] 							= $url;
		$data['skip_asset_check'] = 1;

		if(callURL('POST', $player['address'].'/api/'.$apiVersion.'/assets', $data, $id, false)){
			header('HTTP/1.1 200 OK');
			echo 'Asset added successfully';
			die();
		}
		else {
			header('HTTP/1.1 404 Not Found');
			echo 'Error! - Can \'t add the Asset';
			die();
		}
	}

	if(isset($_POST['changeAssetState'])){
		$playerID 			= $_POST['id'];
		$asset					= $_POST['asset'];
		$playerSQL 			= $db->query("SELECT * FROM player WHERE playerID='".$playerID."'");
    $player 				= $playerSQL->fetchArray(SQLITE3_ASSOC);
		$playerAddress 	= $player['address'];
		$data = callURL('GET', $playerAddress.'/api/'.$apiVersion.'/assets/'.$asset, false, $playerID, false);
		if($data['is_enabled'] == 1 AND $data['is_active'] == 1){
			$data['is_enabled'] = '0';
			$data['is_active'] 	= '0';
		}
		else {
			$data['is_enabled'] = '1';
			$data['is_active'] 	= '1';
		}
		if(callURL('PUT', $playerAddress.'/api/'.$apiVersion.'/assets/'.$asset, $data, $playerID, false)){
			header('HTTP/1.1 200 OK');
			exit();
		} else {
			header('HTTP/1.1 404 Not Found');
			exit();
		}
	}

	if(isset($_POST['changeAsset'])){
		$playerID 			= $_POST['playerID'];
		$orderD 				= $_POST['order'];
		$playerSQL 			= $db->query("SELECT * FROM player WHERE playerID='".$playerID."'");
    $player 				= $playerSQL->fetchArray(SQLITE3_ASSOC);
		$playerAddress 	= $player['address'];
		$result = callURL('GET', $playerAddress.'/api/v1/assets/control/'.$orderD.'', false, $playerID, false);
		$db->exec("UPDATE `player` SET sync='".time()."' WHERE playerID='".$playerID."'");
		if($result != ''){
			header('HTTP/1.1 200 OK');
			echo $result;
		}
		else header('HTTP/1.1 404 Not Found');
	}

	if(isset($_POST['exec_reboot'])){
		$playerID 			= $_POST['playerID'];
		$playerSQL 			= $db->query("SELECT * FROM player WHERE playerID='".$playerID."'");
		$player 				= $playerSQL->fetchArray(SQLITE3_ASSOC);
		$playerAddress 	= $player['address'];
		$db->exec("UPDATE `player` SET sync='".time()."' WHERE playerID='".$playerID."'");
		header('HTTP/1.1 200 OK');
		echo 'Reboot command send!';
		$result = callURL('POST', $playerAddress.'/api/v1/reboot_screenly', false, $playerID, false);
	}

	if(isset($_POST['editInformation'])){
		$playerID 	= $_POST['playerID'];
    $playerSQL 	= $db->query("SELECT * FROM player WHERE playerID='".$playerID."'");
    $player 		= $playerSQL->fetchArray(SQLITE3_ASSOC);
		if($playerID != ''){
			header('HTTP/1.1 200 OK');
			$return_arr = array("player_name" => $player['name'], "player_address" => $player['address'], "player_location" => $player['location'], "player_user" => $player['player_user'], "player_password" => $player['player_password']);
			echo json_encode($return_arr);
		}
		else header('HTTP/1.1 404 Not Found');
	}
?>
